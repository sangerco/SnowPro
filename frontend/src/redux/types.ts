export interface UserData {
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
  }


export const FETCH_USER_DATA_REQUEST = 'FETCH_USER_DATA_REQUEST';
export const FETCH_USER_DATA_SUCCESS = 'FETCH_USER_DATA_SUCCESS';
export const FETCH_USER_DATA_FAILURE = 'FETCH_USER_DATA_FAILURE';


export interface FetchUserDataRequestAction {
    type: typeof FETCH_USER_DATA_REQUEST
}

export interface FetchUserDataSuccessAction {
    type: typeof FETCH_USER_DATA_SUCCESS;
    payload: UserData;
}

export interface FetchUserDataFailureAction {
    type: typeof FETCH_USER_DATA_FAILURE;
    payload: string;
}

export type UserActionTypes = 
    FetchUserDataRequestAction
    | FetchUserDataSuccessAction
    | FetchUserDataFailureAction;