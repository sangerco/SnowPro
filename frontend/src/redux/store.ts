import { Action, compose } from "redux";
import { ThunkAction, configureStore } from "@reduxjs/toolkit";
import { logger } from "redux-logger";
import rootReducer from "./reducers/rootReducer";

const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(logger, composeEnhancers),
});
