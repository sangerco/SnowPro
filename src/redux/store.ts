import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
