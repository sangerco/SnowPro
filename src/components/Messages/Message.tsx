import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  Button,
  Dimmer,
  Loader,
  Icon,
  Card,
  Modal,
  Grid,
  Header,
  Divider,
} from "semantic-ui-react";
import { Link, useParams } from "react-router-dom";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchMessage, deleteMessage } from "../../redux/slices/messageSlice";
import MessageReplyView from "./MessageReplyView";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";

const Message: React.FC = () => {
  const { id } = useParams();
  const messages = useSelector((state: RootState) => state.messages);
  const dispatch = useDispatch<AppDispatch>();

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  useEffect(() => {
    if (id) {
      dispatch(fetchMessage(id));
    }
  }, [dispatch, id]);

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  const handleShowDeleteModal = () => {
    setShowDeleteModal(true);
  };

  const handleDeleteMessage = () => {
    if (id) {
      dispatch(deleteMessage(id));
    }
  };

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  if (messages.loading) {
    return (
      <div>
        <Dimmer active>
          <Loader>Loading...</Loader>
        </Dimmer>
      </div>
    );
  }

  if (messages.error) {
    return <div>Error: {messages.error}</div>;
  }

  if (Array.isArray(messages.message)) {
    return <div>Error: Message cannot be loaded!</div>;
  } else if (messages.message) {
    const message = messages.message;

    return (
      <Grid>
        <Grid.Row>
          <Grid.Column width={4}>
            {skiAreaState.loading ? (
              <Dimmer active>
                <Loader>Loading...</Loader>
              </Dimmer>
            ) : skiAreas && skiAreas.length > 0 ? (
              <Card style={{ marginTop: "10px", marginLeft: "20px" }}>
                {skiAreas.map((sa) => (
                  <Card.Content key={sa.slug}>
                    <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                  </Card.Content>
                ))}
              </Card>
            ) : skiAreaState.error ? (
              <Dimmer active>
                <Header as="h1">
                  Error! Ski Area Data cannot be retrieved! {skiAreaState.error}
                </Header>
              </Dimmer>
            ) : null}
          </Grid.Column>
          <Grid.Column width={8}>
            <Card fluid>
              <Card.Content>
                <Card.Header>{message.subject}</Card.Header>
                <Divider />
                <Card.Content>
                  {message.senderFirstName} {message.senderLastName}
                </Card.Content>
                <Card.Meta textAlign="right">
                  {formatDate(message.createdAt)}
                </Card.Meta>
                <Divider />
                <Card.Description>{message.body}</Card.Description>
              </Card.Content>
              <Card.Content extra>
                <div className="ui two buttons">
                  <Button basic color="red" onClick={handleShowDeleteModal}>
                    <Icon name="trash" />
                  </Button>
                  <Button basic color="green">
                    <Link
                      to={`/messages/${message.id}/reply`}
                      style={{ color: "green" }}>
                      Reply
                    </Link>
                  </Button>
                </div>
              </Card.Content>
            </Card>
            {message.replies && message.replies.length > 0
              ? message.replies.map((reply) => (
                  <MessageReplyView key={reply.id} messageReply={reply} />
                ))
              : null}
            <Modal
              open={showDeleteModal}
              onClose={() => setShowDeleteModal(false)}>
              <Modal.Content>
                Are You Sure You Want To Delete This Message?
              </Modal.Content>
              <Modal.Actions>
                <Button negative onClick={handleDeleteMessage}>
                  Yes
                </Button>
                <Button onClick={() => setShowDeleteModal(false)}>
                  Cancel
                </Button>
              </Modal.Actions>
            </Modal>
          </Grid.Column>
          <Grid.Column width={4}></Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }

  return null;
};

export default Message;
