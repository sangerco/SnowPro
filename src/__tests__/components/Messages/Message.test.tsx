/* eslint-disable testing-library/no-node-access */
import React from "react";
import Message from "../../../components/Messages/Message";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import { BrowserRouter } from "react-router-dom";
import nock from "nock";

describe("test Message", () => {
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Message />
        </BrowserRouter>
      </Provider>
    );
  });

  it("matches a snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Message />
        </BrowserRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("testing message view", () => {
  const id = "message-1";
  const useParamsMock = jest.requireMock("react-router-dom")
    .useParams as jest.Mock;
  useParamsMock.mockReturnValue({ id });
  it("should show a message", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get(`/messages/${id}`)
      .reply(200, {
        message: {
          id: "message-1",
          senderId: "testuser",
          recipientId: "testuser2",
          subject: "test message",
          body: "test message body",
          createdAt: new Date(),
          isRead: false,
          senderUsername: "testuser",
          senderFirstName: "Test",
          senderLastName: "User",
          recipientUsername: "testuser2",
          recipientFirstName: "Test",
          recipientLastName: "User2",
          replies: [],
        },
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Message />
        </BrowserRouter>
      </Provider>
    );

    // eslint-disable-next-line testing-library/no-debugging-utils
    screen.debug();

    expect(nock.pendingMocks().length).toEqual(2);

    await waitFor(() => {
      const message = screen.getByTestId("message");
      expect(message).toBeInTheDocument();
    });

    await waitFor(() => {
      const deleteButton = screen.getByTestId("delete-button");
      expect(deleteButton).toBeInTheDocument();
    });

    await waitFor(() => {
      const replyButton = screen.getByTestId("reply-button");
      expect(replyButton).toBeInTheDocument();
    });
  });

  it("should display the delete modal", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get(`/messages/${id}`)
      .reply(200, {
        message: {
          id: "message-1",
          senderId: "testuser",
          recipientId: "testuser2",
          subject: "test message",
          body: "test message body",
          createdAt: new Date(),
          isRead: false,
          senderUsername: "testuser",
          senderFirstName: "Test",
          senderLastName: "User",
          recipientUsername: "testuser2",
          recipientFirstName: "Test",
          recipientLastName: "User2",
          replies: [],
        },
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Message />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const deleteButton = screen.getByTestId("delete-button");
      // eslint-disable-next-line testing-library/no-wait-for-side-effects
      fireEvent.click(deleteButton);
    });

    await waitFor(() => {
      const deleteModal = screen.getByTestId("delete-modal");
      expect(deleteModal).toBeInTheDocument();
    });
  });
});
