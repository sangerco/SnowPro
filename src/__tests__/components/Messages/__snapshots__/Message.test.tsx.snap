// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Message matches a snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="ui active transition visible dimmer"
      style="display: flex;"
    >
      <div
        class="content"
      >
        <div
          class="ui text loader"
        >
          Loading...
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
